### ISSUE TITLE
SageAttention2 will support qo len != kv len with causal?

### ISSUE THREAD

**Issue by yongchaoding** [thu-ml/SageAttention#131]
I want to use SageAttention2 into LLM applications, and in common case, LLM will cache some prompt, but not all. so at this state, qo len != kv len in Q and KV Cache, and then SageAttention2 will has error result with or without causal set. So I wonder if SageAttention will support it soon?

**Comment by jason-huang03** [comment]
@yongchaoding we originally plan to support this. However, we were not certain about how this should behave (e.g. which token should be masked). Can you point us to some reference?

**Comment by Z-Z188** [comment]
> [@yongchaoding](https://github.com/yongchaoding) we originally plan to support this. However, we were not certain about how this should behave (e.g. which token should be masked). Can you point us to some reference?

作者您好，我认为他说的可能是下列的情况：
sageattention和flash_attn_with_kvcache有很大的误差，在流式视频生成中无法正确生成。
请问这是为什么呢

```python
from sageattention import sageattn
from flash_attn import flash_attn_interface
import torch

B = 4
N = 1560
H = 12
D = 128

device = "cuda:0"
dtype = torch.float16

k_cache = torch.randn(B, 3*N, H, D, device=device, dtype=dtype)
v_cache = torch.randn(B, 3*N, H, D, device=device, dtype=dtype)
q = torch.randn(B, N, H, D, device=device, dtype=dtype)
effective_seqlens = torch.tensor([1560, 3120, 4680, 4680], device=device, dtype=torch.int32)


x = flash_attn_interface.flash_attn_with_kvcache(
    q=q,
    k_cache=k_cache,
    v_cache=v_cache,
    cache_seqlens=effective_seqlens,
)

outputs = []
for b in range(q.shape[0]):
    seqlen = effective_seqlens[b]

    k = k_cache[b:b+1, :seqlen]
    v = v_cache[b:b+1, :seqlen]
    print(f"k shape: {k.shape}")

    out = sageattn(
        q[b:b+1],
        k,
        v,
        tensor_layout="NHD",
        is_causal=False,
    )
    outputs.append(out)

x_ret = torch.cat(outputs, dim=0)

print(f"flash-attn: {x.mean()}")
print(f"sageattn: {x_ret.mean()}")
```
